content/advent-2019/testing-buffalo-apps.md

```diff
@@ -0,0 +1,394 @@
+       actionAccessed = true
+       return nil
+   }
+)
+
+func (ms *MiddlewareSuite) Test_Company_Middleware() {
+   ms.LoadFixture("Load companies")
+   app := buffalo.New(buffalo.Options{})
+   app.Use(popmw.Transaction(models.DB))
+ 
+   actionAccessed = false
+   g := app.Group("/{company_id}/")
+   g.Use(AuthorizeCompany)
+   g.GET("/sample", sampleHandler)
+ 
+   ht := httptest.New(app)
+   res := ht.HTML("/%v/sample", 111).Get()
+   ms.Equal(false, actionAccessed)
+   ms.Equal(http.StatusNotFound, res.Code)
+   
+  // Look for loaded company in the DB
+   company := models.Company{}
+   err := models.DB.Last(&company)
+   ms.NoError(err)
+ 
+   res = ht.HTML("/%v/sample", company.ID).Get()
+   ms.Equal(true, actionAccessed)
+   ms.Equal(http.StatusOK, res.Code)
+}
+
+```
+
+
+## Helpers 
+
+Helpers are functions we use in plush views that help us abstract things that we don’t want or cannot do in the view layer. A Buffalo helper function looks like:
+
+```go
+//HelperFormatDate is a helper function intended to be available globally to simplify date formatting on templates
+func HelperFormatDate(t time.Time) string {
+   date := t.Format("01/02/06")
+   return date
+}
+```
+
+And are added to buffalo apps on the render.go `init()` function. 
+
+```go
+func init() {
+   r = render.New(render.Options{
+       // HTML layout to be used for all HTML requests:
+       HTMLLayout: "application.plush.html",
+ 
+       // Box containing all of the templates:
+       TemplatesBox: templatesBox,
+       AssetsBox:    assetsBox,
+ 
+       // Add template helpers here:
+       Helpers: render.Helpers{
+           // for non-bootstrap form helpers uncomment the lines
+           // below and import "github.com/gobuffalo/helpers/forms"
+           // forms.FormKey:     forms.Form,
+           // forms.FormForKey:  forms.FormFor,
+           "formatDate": HelperFormatDate,
+       },
+   })
+}
+```
+
+In the (you guessed it) Helpers property of the rendering engine (r) created there.
+
+As the helpers are pure functions that take input and outputs you can just test these functions separately. For example:
+
+```go
+import (
+   "time"
+)
+ 
+func (as *ActionSuite) Test_FormatDateHelper() {
+   date := "2019-12-31"
+   t, _ := time.Parse("2006-01-02", date)
+ 
+   formattedDate := HelperFormatDate(t)
+   as.Equal("12/31/19", formattedDate)
+}
+```
+
+This test ensures that the resulting date is formatted accordingly.
+
+## Actions
+
+Actions are the C in the MVC design pattern is where we join our application domain and uses the view to represent a correct response to the user (or external system in the case of an API).
+
+These are also called handlers. And its tests are typically integration tests. This is because Actions are not intended to store business logic in them (Fat controller anti-pattern), hence testing Actions will imply using the model and the view layers to test the complete operation. 
```

Comment:
```suggestion
These are also called handlers, and its tests are typically integration tests. This is because Actions are not intended to store business logic in them (Fat controller anti-pattern), hence testing Actions will imply using the model and the view layers to test the complete operation. 
```